core/src/main/java/com/linecorp/armeria/common/RequestCancellationException.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.common;
+
+/**
+ * A {@link RuntimeException} raised when a request is cancelled by the user.
+ */
+public final class RequestCancellationException extends RuntimeException {
```

Comment:
Should we make `TimeoutException` extend this class?

Reply:
Then `ResponseTimeoutException` will be an instance of `RequestCancellationException` 🤔. That's a bit mixed up.

Reply:
Yeah, that's a little bit weird. Let's leave it as it is for now. Could you add some TODO about this as an inline comment? e.g.

    TODO(trustin): Clean up the type hierarchy between RequestCancellationException,
                   TimeoutException, RequestTimeoutException and ResponseTimeoutException.

Reply:
> Question: Don't we need to introduce `ResponseCancellantionExeption` for consistency?

I guess that will make the type hierarchy unexpressible in Java. How about merging `TimeoutException` and `ResponseTimeoutException` into `RequestTimeoutException` instead?

Reply:
For example, we could:

1. Deprecate `TimeoutException` and `ResponseTimeoutException`;
2. Make `ResponseTimeoutException` extend `RequestTimeoutException`; and
3. Make `TimeoutException` extend `RequestCancellationException`

However, we'll have to:

- Document this change carefully in the changelog; and
- Maybe pass `RequestContext` of the cancelled request so that a user can tell if the timeout is from which side.

I think this is not the scope of this pull request, though. We can follow up later. Thoughts?

Reply:
> Question: Don't we need to introduce `ResponseCancellantionExeption` for consistency?

Yeah, `RequestCancellationException` and `ResponseCancellationException` could extend a new parent class `CancellationException` which is similar to `TimeoutException`.

Reply:
Could the type hierarchy like the following?

```
class CancellationException
class RequestCancellationException extends CancellationException
class ResponseCancellationException extends CancellationException

class TimeoutException extends CancellationException
class RequestTimeoutException extends TimeoutException
class ResponseTimeoutException extends TimeoutException
```

Reply:
@ikhoon @minwoox @tumile Thoughts?

Reply:
LGTM. 👍

Reply:
+1